### DISCUSSION TITLE
PID output to actuator control

### DISCUSSION THREAD

**Discussion by batman11111** [jeff-burright/Autopilot_ESP32_wifi#8] [General]
Hi Jeff, might be a dumb question but here it is anyway. In my setup I do not have a rudder position sensor and Im using a linear actuator to move the rudder left and right at a constant speed. Based on the compass error how does the PID system know when to move the rudder left (for example) to correct the course and then how does it know when to pull it back to center to avoid an over correction? 

For example: lets say the boat is 10 degrees off course to the left. The tiller would have to move left and then as the error is reduced it would have to move right (back to center) or else it would severely over correct if the tiller just stays out to the left. If I understand correctly the PID output would be something like this:

t= 0 seconds      error= -10 degrees     PID output: -270     (rudder is moving left now)
t= 1 second        error= -9 degrees      PID output: -200      (error is decreasing, PID output is decreasing but the rudder is still moving left since the actuator is moving at a static speed and the only options are left and right, negative for left, positive for right)
.
.
.
t=10 seconds      error -1 degrees       PID output:  -10  (all PID output is negative so the rudder would only move to the left the entire time causing a major over correction)

Would I have to add in a line of code that says move the rudder out to the left until the error is no longer increasing and then use the PID output to bring the rudder back to center?

Im sure Im just missing something but I cannot figure out what it is. Can you help point me in the right direction (pun intended) Thank you!

**Comment by jeff-burright** [comment]
Hi, not a dumb question and one I'm having to stretch my memory to remember and explain correctly. First, the rudder position sensor is really only there for a wheel-based pilot using a motor without end stops. It allows you to preset an angle of the sensor at which it'll send a stop command to the motor so you don't jam against the steering limit of the wheel. I think there's a way to simulate this with the voltage sensing abilities of the IBT-2 motor controller, watching for a surge when the motor hits its limit, but I never devoted the time to make that work for me. The linear actuator I got off ebay has end stops in it, so you don't need a rudder sensor. 

Your other question relates to the "D" in PID and the bearingrate term (calculated in the Subs.ino tab). Here is an explanation from Jack's original documentation that he sent to me when I started this project:

"Undersatanding the PID

The PID is the math part that calculates how much rudder is needed.  The PID most people are familiar with is the cruise control in their car.  When you turn on your cruise control you give the cruise control a setpoint. The PID has three parts.  The first and most important is the Proportional part. Read about PIDs on the internet.

Heading Error = Current Heading – Heading Setpoint. 

Rudder Command(Proportional) = K x Heading Error. 

 For K I am using .8, therefore if I am 10 degrees off course I will get a command for 8 degrees of rudder.

0.8 x 10 = 8 

That is fine now you start to turn back to course.  As the heading error goes to zero I am back on course and the rudder goes back to zero.  But at that point I am still turning, I need to counter steer.

That is where the D in PID comes in.  The D stands for differential and it is how fast I am turning or bearing rate.  So if I am turning I want to subtract from the rudder command based on how fast I am turning.

Putting this together I have.

Rudder Command = Rudder Command Proportional   - Rudder Command Differential

Kp x Heading Error – Kd x Bearing Rate.

That’s it,  That what controls the boat.

The I in PID stands for Integral but it is not used.

Bearing rate is a measure of how fast the boat is turning in degrees per second, and when the boat is not turning it should be zero.  Bearing rate is used in the PID to stabilize the autopilot and prevent overshoot."

Does that help?

**Reply by batman11111** [reply]
Somewhat, I understand everything you wrote except the following. 

You wrote: "For K I am using .8, therefore if I am 10 degrees off course I will get a command for 8 degrees of rudder.
0.8 x 10 = 8"  

This is the part that isn't connecting for me yet. Without a rudder position sensor, how would I know if I have 8 degrees of rudder or not? Or is this somehow tied into whatever the compass heading is? Thanks for responding so quick! I've been trying to figure this out for about 2 weeks now.

**Reply by jeff-burright** [reply]
Ah, I think maybe I see. When it says 10 degrees rudder, what is really meant is there is a 10-degree heading error (e.g., my course to steer is 270 degrees, but the compass says my heading is 260, so my heading error is 10 degrees). This is purely the difference between the compass reading and the user-defined desired course. The PID function will then cause the motor controller to move the actuator to turn the boat to the right. At each timestep the heading error will be less and less (reducing the "P" term), further mediated by the speed of the turn (bearing rate based on IMU measurements), which increases the "D" term until it overpowers the shrinking "P" and tells the actuator to *reverse* (negative value) as you approach the desired heading. The PID function essentially "swims" toward your desired heading, using only compass measurements, with a goal of centering the tiller over enough course corrections. By tuning the P and D settings (Jack referred to the "P" as "K" in his explanation), you can find the balance between understeering and oversteering to stay inside the deadband of 2 degrees (or whatever you set it as). 

That's how I understand it anyway. There is a version of an autopilot where you could dictate what angle to hold the rudder for any given combination of heading error/bearing rate, but it would still follow the basic PID course correction pattern. Jack's original code had a rudder position-based PID function that offers more precision, but it is purely optional. 

In the PID.ino tab, you will see starting on line 62 that the output of the PID function is "rudder_command". This is then converted to "rudder_error" on line 139. If you have a rudder position sensor, the rudder_error would be transformed by knowing where the rudder position currently is, but otherwise the rudder position is not accounted for at all in the function (rudder_position = 0). Then starting on line 186 you will see that if the rudder error is positive, you go right, and if it's negative, you go left. So, when D overpowers P, it changes the direction of the actuator. The term "motorspeed" dictates how fast to move the actuator based on  rudder_error. Hence, farther off course => faster rate of turn => shrinking P/growing D => buffer against fishtail-oversteering.

**Reply by jeff-burright** [reply]
PS - I edited the above a bunch of times to try to make it clearer, so you might want to reload to make sure you're reading the latest and greatest.

**Reply by batman11111** [reply]
This is great! I was missing the part where D is tied to angular velocity and goes negative the more the boat corrects, hence bringing the rudder back to center before overshooting and I can tune the P and D based on the turning characteristics of my boat. Thanks so much for your help, this has stumped me for a couple weeks now.
